### ISSUE TITLE
makefile: add more targets (especially to run all tests and remove -Werror)

### ISSUE THREAD

**Issue by braoult** [exercism/c#687]
For me, it is very difficult to use the current C-track `makefile`, because :
1) it is quite rigid (it overrides CFLAGS at first) - i.e. it must be changed to add any flag
2) there are missing targets, which could make life much easier:
* possible new target 1: `testall`
This one would be to run all tests. I did implement this by adding a few lines in my exercises include files :
```
#ifdef  TESTALL
#undef  TEST_IGNORE
#define TEST_IGNORE() {}
#endif
```
I believe there must be easier ways to do it, but I don't know the test suite enough and I did choose a _lazy_ solution. Then we can simply pass `-DTESTALL` to compiler for this target... Note: This code should be moved to test source code, if it is accepted.

* possible new target 2: `noerror`
This one would be to allow a source with warnings to compile. Simple example :
```
switch (a) {
   case 1:
     do_something;
   case 2:
     do_something_else;
     break;
}
```
This code fails with our current `CFLAGS` with `GCC`, but will perfectly compile with `CLANG`.

Warnings (especially `-Wall`, `-Wextra`) are heavily compiler dependent, and failing with one compiler and not with another one looks weird : imagine it compiles on student side, and not on mentor side, it will be difficult for mentor to 1) check that warnings can be ignored,  2) edit `makefile` to remove `-Werror`, and 3) build again. The mentor could even (maybe wrongly) ask for a resubmit.

What we need (for me) is to avoid as much as possible any manual files change, when it can be done by an additional `make` target.

I did raise the issue on [Slack 2 days ago](https://app.slack.com/client/TAN6QMALR/CATCU7LSJ/thread/CATCU7LSJ-1630094279.041600). Note: I think Slack does not keep discussions forever, maybe we should copy the thread here ?

We could later imagine more targets, for example one for `memcheck+testall`, one for `testall+noerror`, etc...

My own implementation (GNU make only) uses a `GNUmakefile`, which includes `makefile` and adds/removes some flags depending on targets. If we could include the equivalent in default  `makefile`, I think it could save a lot of time, both for students and mentors.

Example: [My `GNUmakefile`](https://github.com/braoult/exercism/blob/master/c/templates/GNUmakefile). You will notice it has also some targets to build standalone exercises (without test suite code). I not sure people will want this, I simply find much easier to call the exercise from my own `main()` than editing the test source code, when I want to test some case not covered by the exercise test suite.

Does it make any sense ?

**Comment by braoult** [comment]
Working example with my `GNUmakefile` :

* `make` with `GCC` : fail
```
br@lorien:.../exercism/c/phone-number$ CC=gcc make
rm -rf *.o *.out *.out.dSYM
Compiling tests.out
src/phone_number.c: In function ‘phone_number_clean’:
src/phone_number.c:50:14: error: this statement may fall through [-Werror=implicit-fallthrough=]
   50 |             p++;
      |             ~^~
src/phone_number.c:52:9: note: here
   52 |         case 3:                                   /* last 3 numbers */
      |         ^~~~
cc1: all warnings being treated as errors
make: *** [makefile:37: tests.out] Error 1
```
* `make` with `CLANG` : success
```
br@lorien:.../exercism/c/phone-number$ CC=clang make
rm -rf *.o *.out *.out.dSYM
Compiling tests.out
test/test_phone_number.c:223:test_cleans_the_number:PASS
test/test_phone_number.c:224:test_cleans_numbers_with_dots:PASS
[...]
test/test_phone_number.c:241:test_invalid_if_exchange_code_starts_with_0_on_valid_11_digit_number:PASS
test/test_phone_number.c:243:test_invalid_if_exchange_code_starts_with_1_on_valid_11_digit_number:PASS

-----------------------
18 Tests 0 Failures 0 Ignored 
OK
```
* `make nowarn` with `GCC`: success.
```
br@lorien:.../exercism/c/phone-number$ CC=gcc make nowarn
rm -rf *.o *.out *.out.dSYM
Compiling tests.out
src/phone_number.c: In function ‘phone_number_clean’:
src/phone_number.c:50:14: warning: this statement may fall through [-Wimplicit-fallthrough=]
   50 |             p++;
      |             ~^~
src/phone_number.c:52:9: note: here
   52 |         case 3:                                   /* last 3 numbers */
      |         ^~~~
test/test_phone_number.c:223:test_cleans_the_number:PASS
test/test_phone_number.c:224:test_cleans_numbers_with_dots:PASS
[...]
test/test_phone_number.c:240:test_invalid_if_exchange_code_starts_with_0_on_valid_11_digit_number:PASS
test/test_phone_number.c:242:test_invalid_if_exchange_code_starts_with_1_on_valid_11_digit_number:PASS

-----------------------
18 Tests 0 Failures 0 Ignored 
OK
```

**Comment by ryanplusplus** [comment]
@braoult I'm concerned about adding a lot of complexity to the makefile since many of the users will be unfamiliar with both C and makefiles. If we add several different targets we increase the overhead for these users.

I've found that in most cases there's a pretty simple solution to addressing a warning. Disabling warnings is a bad habit to learn and I'd personally rather encourage understanding and fixing warnings. For example, in the case given above you can let gcc know that the fallthrough is intentional with a comment that includes "fallthrough" (among other options) or by using a goto to jump to the next case explicitly. Neither of these strategies is particularly obvious, but exercism is designed to be a mentored experience which gives us an opportunity to surface things like this.

**Comment by wolf99** [comment]
I have similar opinuon to @ryanplusplus  on this.
Adding complexity and options getting started exponentially harder for people that are not already familiar C and Makefile, or even may not be familiar with build systems at all.

For the errors, again I concur with @ryanplusplus, we aim to teach fully correct C, not just C that will do. If I'm shipping code to a customer then I want to know it's as correct as possible. If we have taught that warnings are ok, even if only by our omission, then students may not learn any different. If there is different behavior on different compliers then that should be resolved, but resolved such that all compilers fail for warnings; would be good to see a PR for this!

In relation to tests, the workflow expected for the exercises is for a student to get the first test working first and only then move on to the next test. In such a scenario, options to "un-ignore" all tests en masse is counter productive. I understand it may be easier in some cases if you are more experienced or a mentor. I would be interested to know more about the use case here.

**Comment by wolf99** [comment]
I don't login to slack often. 
I do try to respond to any C track related questions when I do though.
Probably the best way to discuss more is on issues here like this one.

**Comment by braoult** [comment]
@wolf99 
> In relation to tests, the workflow expected for the exercises is for a student to get the first test working first and only then move on to the next test. In such a scenario, options to "un-ignore" all tests en masse is counter productive.

Where does it come from ? Nobody will never do that. Even unexperienced ones.

They will not edit the test source 10 times for 10 tests... They will simply activate all tests, probable at same time they read the tests source code for the first time to understand what is expected (as README is useless for that). Then they will code, run all tests, and code again, until all tests works.

Note that bash track explicitly says : `To run all tests, including the ones with `skip` annotations, run: XXXX`. Offering a `testall` target would be the same as bash way.

**Comment by ryanplusplus** [comment]
> They will not edit the test source 10 times for 10 tests... They will simply activate all tests, probable at same time they read the tests source code for the first time to understand what is expected (as README is useless for that). Then they will code, run all tests, and code again, until all tests works.

I uncomment tests one by one. This helps to avoid linker errors that force you to write more code than necessary and helps to minimize the amount of distracting feedback that is dumped in the terminal. Having worked with others, this is not an uncommon workflow. I don't know of any major test framework that does not allow you to ignore tests.

**Comment by braoult** [comment]
> > Where does it come from ? Nobody will never do that. Even unexperienced ones.
> > They will not edit the test source 10 times for 10 tests... They will simply activate all tests, probable at same time they read the tests source code for the first time to understand what is expected (as README is useless for that). Then they will code, run all tests, and code again, until all tests works.
> 
> I uncomment tests one by one. This helps to avoid linker errors that force you to write more code than necessary and helps to minimize the amount of distracting feedback that is dumped in the terminal. Having worked with others, this is not an uncommon workflow. I don't know of any major test framework that does not allow you to ignore tests.
> 
> As far as where it comes from, this is a central tenet of writing software with TDD. You focus on one thing at a time and slowly expand the list of requirements. Exercism encourages the TDD methodology.

Not bash one.

**Comment by braoult** [comment]
@ryanplusplus, @wolf99, I cannot say I totally agree...

Note: This issue is not about suppressing warnings, but about adding 2 more targets in `makefile`: `testall`, which runs all tests, and `nowarn`, which suppresses `-Werror` when we feel it could be better (to avoid dirty hack or dirty code just to please a specific compiler with a specific option). The latter could be useful for mentor for example, after understanding the student code is good enough, and warnings can safely be ignored. The goal is to gives more flexibility, not to make the code less secure.

Let start with `testall` target. This one would allow students (and mentors) to run  all tests without editing test source file. This is by the way my default target, I really think it should be exercism default too - I don't really see the point of running one test only, but it is off-topic). Why do we ask students and mentors to edit, for *each* exercise, in *same* file, the *same* lines, when it is not needed at all ? Adding 3 lines in `makefile` and a few directives in test program could save a lots of time.

Second, the `nowarn` target.

What is warning ? This is a compiler specific message telling you it thinks there is something strange in your code. It means a specific compiler option was set to write a message in a specific situation. That's it.
It does not mean the code is wrong: We want to see warnings we decided to see. But the programmer needs the final word to decide if a warning is legitimate or not, and not a `compiler developer`.
Example: I may well decide to declare somewhere: `int fix_the_possible_buf_overflow_below;`, which would remind me forever I have something to do, until I remove the unused variable. I am unsure if an overflow could occur here, but I cannot answer right now. The code is correct C, I want to see the warning, and I want to compile.`

Let's talk about the specific '-Wimplicit-fallthrough=3' which is included in gcc (since 7.4?) with `-Wextra` and not by clang. Note this is `-Wimplicit-fallthrough=3`, not `0`, `1`, `2`, `4`, `5`. The way they control the fallthrough is just different. Numbers 1-4 allows some comments strings (different for each number) before a case to discard the warning.

Why did gcc include it in `-Wextra`? Because of this bad™ comment hack, which allowed to discard the warning. In other words, they decided to **use C comments as input for C compiler**. Some people, like clang guys, say it is an abomination, and explicitely said "never that". I totally agree with clang people here. So they had the warning available (exercism just did not add it), but no hack to hide it (hence clang would never compile with this warning, with `-Werror`).

@wolf99, to be more clear : gcc is at fault here. They enabled the warning, and at same time broke C to have comments interacting with real C code. clang also did include the warning, for longtime. We (exercism) did not add it in our Makefile (we are at fault: we should add each error, not compiler "set", such as `-Wextra`, which can be anything).

I continue. If we had included `-Wimplicit-fallthrough`, for every compiler, 2 years ago, adding `-Werror` would have had clang simply fail compiling any fallthrough.

And all work-arounds are wrong for me :
1) use `goto labelbelow`. Not for me, sorry (and not in kernel too: they prefer to keep getting warnings instead of doing that).
2) use comments as C code (gcc, after 7.4 ?). Just wrong.
3) use if/else only. We let the compiler to decide which is more readable, not possible for me.

For me, *we want to see the warnings we have asked the compiler to check*, be we want to decide which ones could be errors. I mean using: `-Werror=xxx` only, with xxx being what we consider critical (as developer), for example `strict-prototypes`, `implicit-function-declaration`, etc... The kernel has maybe 15 of these. We should do that, not using `-Wextra` (which is unknown) & `-Werror`.

Said a different way, a compiler is not the final decider. The thing (for me) should be : "1) I ask which warning I want to see, 2) the compiler tells me, and 3) I decide".

Lets continue with `-Wimplicit-fallthrough`, clang has eventually accepted in 2019 to follow gcc's `__attribute` for that warning (pushed by the kernel teams, who need to compile Android with Clang only - Google choice for long time then).

Fine you will say, use this attribute... The Linux kernel has:
```
/* from :
 * https://github.com/torvalds/linux/blob/master/include/linux/compiler_attributes.h
 * around line 200
 */
#if __has_attribute(__fallthrough__)
# define fallthrough                    __attribute__((__fallthrough__))
#else
# define fallthrough                    do {} while (0)  /* fallthrough */
#endif
```
Then we just can use `fallthrough;` just before the following `case:`.

But, wait: My current kernel (5.13.14, that I last compiled around July 20th) `Makefile` has something like that around line 790 :
```
ifdef CONFIG_CC_IS_CLANG
KBUILD_CPPFLAGS += -Qunused-arguments
KBUILD_CFLAGS += -Wno-format-invalid-specifier
KBUILD_CFLAGS += -Wno-gnu
# CLANG uses a _MergedGlobals as optimization, but this breaks modpost, as the
# source of a reference will be _MergedGlobals and not on of the whitelisted names.
# See modpost pattern 2
KBUILD_CFLAGS += -mno-global-merge
else

# Warn about unmarked fall-throughs in switch statement.
# Disabled for clang while comment to attribute conversion happens and
# https://github.com/ClangBuiltLinux/linux/issues/636 is discussed.
KBUILD_CFLAGS += $(call cc-option,-Wimplicit-fallthrough,)
endif
```

First you will notice all compilers are not equal for warnings. And you cannot do anything about it.

Second, you will notice they don't want to enable `-Wimplicit-fallthrough` warning for clang yet. Still buggy. It still does not work good enough, 2 years after they decided to adopt gcc "non buggy" (= not with a "C comment as C code") solution.

Again, my prefered solution is not to hide warnings, but to select *myself* if I want to compile anyway. Example, we could have today (perfectly valid for me)
```
case 1:
   a+=2;
   /* we continue */
case 2:
   [...]
```
We know the programmer wants to continue. If we instruct the compiler to check for implicit-fallthrough, it will find the warning. And fail to compile for no reason, if we have blindly set `-Werror`.
The code is as valid as if we used `/* fallthrough */` with gcc only. No change at all. Using `-Werror` is just not correct for me. We can do it of course in exercism, as default, but we must be able to force compilation.

Want to play a game? I played with clang, and added `-Weverything` to exercism makefile.
Result: exercism test source code has warnings, unity source code has warnings... We cannot use exercism anymore. What is the solution ?

**Comment by wolf99** [comment]
This is outlined in this repos docs/TESTS.md which is linked into the track on the website.

This methodology is one of the tenents of TDD, that when implementing, a programmer starts with a single test of the first part of the spec and writes the simplest test for that, then writes the implementation. Then adds a second test that checks the next small part of the spec, then the implementation. Etc etc. 

I think the C Sharp track does something similar, though it's some time since I've looked at code there.

Not sure if you are not familiar with this methodology or just do not agree with the implementation here. 
If the former, then there are a lot of resources to learn more about TDD online and James Grenning has written a recommended book.

**Comment by ryanplusplus** [comment]
@braoult I don't want to argue about whether a specific warning is useful or not. I personally dislike the implicit fallthrough warning for the reasons that you outlined, but in the name of simplicity and pragmatism we have chosen a sane default of `-Wall` and `-Wextra`. We believe that a learner is better served by following a widely applicable default, and that our time is better using this default instead of bike shedding about the optimal set of warnings.

The makefile is there for an advanced user to make whatever changes they want if they disagree with the defaults that we've chosen. As you feel very strongly about this, I encourage you to do so.

**Comment by braoult** [comment]
@wolf99 , I am both student and (new) mentor. I mean very new mentor. I made a mistake by the way: I did not know how to make all (core) exercises without waiting for previous one, so I accepted to mentor. I understood the practice mode too late :-( Coz' I need to understand exercises before correctly mentor them. I regret that: I will need a few weeks before I can help as I would...

@ryanplusplus , you know I don't need to edit anything, you have probably seen my working solution in 1st post above (including my own main function to allow to easily test more edge cases, which are not in exercism itself).
And again, my point is not to suppress a warning, but to allow compilation (when asked to do so) even if there are warnings, for the reasons I gave above.

Well, this whole issue was opened to give more tools to other students and mentors, to avoid (useless) repetitive work. But if it is not suitable here, this is all good for me...

If you prefer to close this, let me know (I think we will not progress a lot, our views are too different for both `testall` and `nowarn`, for both student and mentor sides).
